[Datastore] Fix ResourceCache table caching key bug

alxtkr77 opened the pull request:
## Summary
- Fix bug in `ResourceCache.get_table()` where `parse_path(uri)` overwrote the `uri` variable
- Tables were being stored under the parsed path instead of the original URI, causing cache misses
- This led to duplicate Table objects being created on subsequent calls with the same URI

## Root Cause
In `store_resources.py`, the code was:
```python
endpoint, uri = parse_path(uri)  # uri gets overwritten!
self._tabels[uri] = Table(uri, ...)  # stored under wrong key
```

The fix preserves the original URI for the cache key:
```python
endpoint, path = parse_path(uri)  # uri preserved
self._tabels[uri] = Table(path, ...)  # stored under correct key
```

## Related Issue
[ML-11518](https://jira.iguazeng.com/browse/ML-11518)

## Test Plan
- Existing unit tests pass
- Verified fix addresses table caching behavior during model monitoring drain cycles

[ML-11518]: https://iguazio.atlassian.net/browse/ML-11518?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
[Datastore] Fix ResourceCache table caching key bug (ML-11518)

Fix bug where parse_path() overwrote the uri variable, causing tables
to be stored under the parsed path instead of the original URI. This
caused cache misses on subsequent get_table() calls with the same URI,
leading to duplicate Table objects being created.

The fix preserves the original URI for the cache key while using the
parsed path for the Table constructor, ensuring proper cache hits.

gtopper reviewed (commented):
Looks great. Would be good to add a regression unit test.

Commit:
[Datastore] Add regression test for table caching fix (ML-11518)

Test verifies that ResourceCache.get_table() caches tables under the
original URI, not the parsed path. This prevents cache misses and
duplicate Table objects on subsequent calls with the same URI.

gtopper commented on tests/datastore/test_datastores.py:
```diff
@@ -380,3 +380,42 @@ def test_item_to_real_path_map(virtual_path: str, tmpdir: Path) -> None:
+    """
+    from unittest.mock import MagicMock, patch
+
+    from mlrun.datastore.store_resources import ResourceCache
```
I don't see any reason to import these locally.

gtopper commented on tests/datastore/test_datastores.py:
```diff
@@ -380,3 +380,42 @@ def test_item_to_real_path_map(virtual_path: str, tmpdir: Path) -> None:
+
+    from mlrun.datastore.store_resources import ResourceCache
+
+    # Arrange
```
```suggestion
```

gtopper commented on tests/datastore/test_datastores.py:
```diff
@@ -380,3 +380,42 @@ def test_item_to_real_path_map(virtual_path: str, tmpdir: Path) -> None:
+            # Act - First call creates the table
+            first_result = cache.get_table(test_uri)
+
+            # Assert - Table created and cached under original URI (not parsed path)
```
These comments are pretty unnecessarily verbose. We know the test will set things up, do something, and make an assertion...

gtopper commented on tests/datastore/test_datastores.py:
```diff
@@ -380,3 +380,42 @@ def test_item_to_real_path_map(virtual_path: str, tmpdir: Path) -> None:
+            assert test_uri in cache._tabels, (
+                f"Table should be cached under original URI '{test_uri}', "
+                f"but cache contains: {list(cache._tabels.keys())}"
+            )
```
This approach is ok, but I think it would be better to test that the table isn't recreated rather than getting into the cache internals. I.e. testing what we care about and not implementation details.

Commit:
[Datastore] Address PR review feedback for table caching test

Refactor test based on reviewer feedback:
- Move imports to file top instead of inline in function
- Remove verbose AAA comments (Arrange/Act/Assert)
- Remove internal cache assertion (_tabels check) in favor of testing
  observable behavior (Table not recreated on second call)

Reference: ML-11518

gtopper commented on tests/datastore/test_datastores.py:
```diff
@@ -380,3 +381,25 @@ def test_item_to_real_path_map(virtual_path: str, tmpdir: Path) -> None:
+
+    Regression test for ML-11518: parse_path() was overwriting the uri variable,
+    causing tables to be stored under the parsed path instead of the original URI.
+    This led to cache misses on subsequent calls with the same URI.
```
We probably don't need an in-depth recap here.

gtopper reviewed (approved):
LGTM otherwise ✅
